api/src/main/java/org/eclipse/microprofile/reactive/messaging/connector/IncomingConnectorFactory.java

```diff
@@ -0,0 +1,108 @@
+ * {@code Incoming} annotation:
+ *
+ * <pre>
+ * &#64;Incoming("my-stream")
+ * public void consume(String s) {
+ *      // ...
+ * }
+ * </pre>
+ * <p>
+ * The set of attributes depend on the connector and transport layer (typically, bootstrap-servers is Kafka specific). The
+ * {@code type} attribute is mandatory and indicates the fully qualified name of the {@link MessagingProvider}
+ * implementation. It must match the class returned by the {@link #type()} method. This is how a reactive messaging
+ * implementation looks for the specific {@link IncomingConnectorFactory} required for a stream. In the previous
+ * configuration, the reactive extension implementation would need to find the {@link IncomingConnectorFactory} returning
+ * the {@code i.e.m.reactive.messaging.impl.kafka.Kafka} class as result to its {@link #type()} method to create the
+ * {@code my-stream} stream. Note that if the connector cannot be found, the deployment must be failed.
+ * <p>
+ * The {@link #getPublisherBuilder(Config)} is called for every stream that needs to be created. The {@link Config} object
+ * passed to the method contains a subset of the global configuration, and with the prefixes removed. So for the previous
+ * configuration, it would be:
+ * <pre>
+ * type =  i.e.m.reactive.messaging.impl.kafka.Kafka
+ * bootstrap-services = localhost:9092
+ * topic = my-topic
+ * </pre>
+ * <p>
+ * So the connector implementation can retrieves the value with {@link Config#getValue(String, Class)} and
+ * {@link Config#getOptionalValue(String, Class)}.
+ * <p>
+ * If the configuration is invalid, the {@link #getPublisherBuilder(Config)} method must throw an
+ * {@link IllegalArgumentException}, caught by the reactive messaging implementation and failing the deployment.
+ * <p>
+ * Note that Reactive Messaging implementation must support the configuration format described here, implementation may
+ * deliver other approaches to create the streams.
+ */
+public interface IncomingConnectorFactory {
+
+    /**
+     * Gets the {@link MessagingProvider} class associated with this {@link IncomingConnectorFactory}.
+     * Note that the {@link MessagingProvider} is a user-facing interface used in the configuration.
+     *
+     * @return the {@link MessagingProvider} associated with this {@link IncomingConnectorFactory}. Must not be
+     * {@code null}. Returning {@code null} triggers a deployment failure.
+     */
+    Class<? extends MessagingProvider> type();
+
+    /**
+     * Creates a <em>stream</em> for the given configuration. The given configuration {@code type} attribute matches the
+     * {@link #type()}.
+     * <p>
+     * Note that the connection to the <em>transport</em> or <em>broker</em> is generally postponed until the
+     * subscription.
+     *
+     * @param config the configuration, never {@code null}
+     * @return the created {@link PublisherBuilder}, must not be {@code null}.
```

Comment:
The 'never' and 'must not be' would be better the other way around as one is a constraint and the other is a promise. So parm 'must not be' and result 'will not be' would read best.